Adding support for taskSpec with pipelineSpec

pritidesai opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Now, this is possible:

```
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: pipelinerun-echo-greetings
spec:
  pipelineSpec:
    tasks:
      - name: echo-good-morning
        taskSpec:
```

Closes #872 

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

Allow embedding task specification with `taskSpec` under `pipelineSpec` such as:

```
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: pipelinerun-echo-greetings
spec:
  pipelineSpec:
    tasks:
      - name: echo-good-morning
        taskSpec:
        ...
```

pritidesai commented:
/assign @vdemeester

pritidesai commented:
@vdemeester please let me know if there are any changes needed ...

dibyom commented on pkg/apis/pipeline/v1alpha1/pipeline_validation.go:
```diff
@@ -126,19 +126,35 @@ func (ps *PipelineSpec) Validate(ctx context.Context) *apis.FieldError {
-			return apis.ErrInvalidValue(strings.Join(errSlice, ","), fmt.Sprintf("spec.tasks[%d].taskRef.name", i))
+		// Check that one of TaskRef and TaskSpec is present
+		if (t.TaskRef == nil || (t.TaskRef != nil && t.TaskRef.Name == "")) && t.TaskSpec == nil {
+			return apis.ErrMissingField(fmt.Sprintf("spec.tasks[%d].taskRef", i), fmt.Sprintf("spec.tasks[%d].taskSpec", i))
```
Nit: We could use `apis.ErrMissingOneOf` for when exactly one of two fields are needed

dibyom commented on pkg/apis/pipeline/v1alpha1/pipeline_validation.go:
```diff
@@ -126,19 +126,35 @@ func (ps *PipelineSpec) Validate(ctx context.Context) *apis.FieldError {
-			return apis.ErrInvalidValue(strings.Join(errSlice, ","), fmt.Sprintf("spec.tasks[%d].name", i))
+		// can't have both taskRef and taskSpec at the same time
+		if (t.TaskRef != nil && t.TaskRef.Name != "") && t.TaskSpec != nil {
+			return apis.ErrDisallowedFields(fmt.Sprintf("spec.tasks[%d].taskRef", i), fmt.Sprintf("spec.tasks[%d].taskSpec", i))
```
Could use `apis.ErrMultipleOneOf` instead of `ErrDisallowedFields`

dibyom commented on Gopkg.lock:
```diff
@@ -1370,6 +1370,7 @@
     "go.uber.org/zap/zaptest",
     "go.uber.org/zap/zaptest/observer",
     "golang.org/x/xerrors",
+    "k8s.io/api/apps/v1",
```
Hmm...I don't see this new import being used anywhere in this commit. Is this needed?

pritidesai replied:
humm, not sure, this is autogenerated ...

pritidesai replied:
this disappeared after rebasing with master 🤔

dibyom commented on pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:
```diff
@@ -245,20 +245,29 @@ func ResolvePipelineRun(
 		// Find the Task that this PipelineTask is using
 		var t v1alpha1.TaskInterface
 		var err error
```
Could be grouped in a var declaration block:
```
var(
  t v1alpha1.TaskInterface
  err error
  ...
)

pritidesai commented:
/retest

pritidesai commented:
> Thanks @pritidesai ...A few small nits. Otherwise looks good!
> Also would you mind updating the Release Notes section of the PR description?

Added same example to be clear and adding more value in Release Notes, would that be ok?

pritidesai commented:
/test pull-tekton-pipeline-build-tests

pritidesai commented:
/retest

pritidesai commented:
Thanks @dibyom for the review, I have addressed all of your feedback/comments, its building as well (after multiple tries :( ) ...

afrittoli commented on pkg/apis/pipeline/v1alpha1/pipeline_validation.go:
```diff
@@ -128,19 +128,35 @@ func (ps *PipelineSpec) Validate(ctx context.Context) *apis.FieldError {
+		if t.TaskSpec != nil {
+			if err := t.TaskSpec.Validate(ctx); err != nil {
+				return err
+			}
```
NIT: I guess we could return nil here, since we have a spec so we cannot have a TaskRef and we checked for that...

pritidesai replied:
line 142 returns an error from the taskSpec validation, if a task has `taskSpec`, validate the specification and return error if validation fails.

vdemeester replied:
@pritidesai it's a nit but I tend to be on @afrittoli side in that, if the `TaskSpec` is valid, we can safely `return nil` and do not care about the rest of the function. As of the code is done today, the rest of the function will be no-op, but I tend to prefer returning early and explicitely

pritidesai replied:
haha I am confused now 😢 `return err` on line 142 returns `error` if `taskSpec` validation fails otherwise continues to next task in a `for` loop. you saying `return  err` could be replaced with `return nil` 🤔 or add `return nil` after line 144?

vdemeester replied:
@pritidesai sorry about that :stuck_out_tongue: Also, just realized it's in a loop, so the `return nil` wouldn't be correct :stuck_out_tongue:. We could extract that `validate` part on its own, but we can do that later on :wink: so, forget I ever said anything :bowing_man: :joy_cat: :stuck_out_tongue_closed_eyes:

afrittoli replied:
Oh, yeah, my bad, sorry about the confusion :(

pritidesai replied:
no worries @afrittoli 👍

pritidesai commented on examples/pipelineruns/pipelinerun-with-pipelinespec-and-taskspec.yaml:
```diff
@@ -0,0 +1,32 @@
+              command:
+                - echo
+              args:
+                - "$(inputs.params.MESSAGE)"
```
hey @afrittoli `$(params.GREETINGS)` will not work here as task is aware of only one input param named `MESSAGE`, here in this example, `GREETINGS`'s value is assigned to task's input param `MESSAGE`. If we replace line 26 with `$(params.GREETINGS)`, pod will have a output of:

```
$(params.GREETINGS)
```

vdemeester replied:
I wonder if it should work or not though… My gut feeling as a user would be that I can use pipeline params in the embedded `TaskSpec` (which is not the case with this PR), right ?

sbwsg replied:
I guess if we wanted that we'd need syntax like `pipeline.params.FOO`, right?  Since both the pipeline and the task could have a param called FOO and they might have different intended meanings.  So `$(params.FOO)` would be a bit confusing (to Tekton controller, but also to users).  Or we could assume that task params always "win" if there's a same named param in both pipeline and task.

vdemeester replied:
> I guess if we wanted that we'd need syntax like `pipeline.params.FOO`, right? Since both the pipeline and the task could have a param called FOO and they might have different intended meanings. So `$(params.FOO)` would be a bit confusing (to Tekton controller, but also to users). Or we could assume that task params always "win" if there's a same named param in both pipeline and task.

Not entirely sure to be honest. I think the main question is, should the embedded `TaskSpec` in a `PipelineSpec` have the exact same feature fields as a normal `TaskSpec` ? Should we have `Params` at all in the embedded `TaskSpec` at all ? (from a user perspective, we don't really need to care about the implementation). 

As a user, if I use an embedded `TaskSpec`, I kinda expect that the `PipelineSpec.params` can be used "as is" in the rest of the struct, which is not the case with the current state. I'm wondering if this is what others would expect or not :angel:

sbwsg replied:
oh i see.  Yeah I like that.  Pipeline has params but embedded task doesnt need them and just defers to the pipeline.

Implementation of that could be quite simple - embedded task spec is "hydrated" with a copy of the params from the pipeline before the taskrun is started.

pritidesai replied:
added couple of more variations just for users to understand it until we change the existing behavior.

pritidesai commented:
> Thank you for this, very useful.
> I have a couple of minor comments / questions - let me know what you think.

Thanks @afrittoli, I have addressed your comments and questions. Let me know what you think.

afrittoli commented:
NIT: In case you have to respin respin this for the test failures, it would be nice to document that the embedded task cannot refer do the PipelineRun parameters directly. we can do that as follow up too.
Thanks!
/lgtm

pritidesai commented:
> NIT: In case you have to respin respin this for the test failures, it would be nice to document that the embedded task cannot refer do the PipelineRun parameters directly. we can do that as follow up too.
> Thanks!
> /lgtm

thanks @afrittoli hopefully have fixed the build failure, had not pulled latest changes from master 😢

pritidesai commented:
/retest

pritidesai commented:
Its ready for review again 🤔

afrittoli commented:
Thank you for the updates! Would you mind filing an issue about supporting access to pipeline params directly in the task spec, so we don't forget about it? I can also create it if you prefer so.
/lgtm

pritidesai commented:
Hey @afrittoli just opened an issue here #1879, please correct it if my interpretation/understanding is off, also sorry for delay, not feeling well past few days, my twin boys passed on me something they have been sick with 😢

pritidesai commented:
/test pull-tekton-pipeline-build-tests

pritidesai commented:
@afrittoli  its ready for review again, had to resolve conflicts 😭

afrittoli commented:
Thank you for this!
/lgtm
